Improve bmp regression performance

ElliotFreebairn opened the pull request:
This the flame graph pre changes

<img width="220" height="174" alt="image" src="https://github.com/user-attachments/assets/d87f84f0-76f6-4a5b-94c0-9d15edaec907" />

This is just a work in progress change, but In the flame graph `recalculate_masks` contributes to a significant amount of CPU usage. My thought process is that, we can sacrifice storing them as member variables and thus are not forced to recalculate them every time set_data is called. It worked quite well, reducing the execution time on a run from 19 seconds to 13 seconds.

However, It has lead to significantly more code and repetition especially in `main`/tests. So I was wondering, is the trade-off worth it and/or a better way to go about it.

I've got other ideas on improving performance, which I'll starting working on them by the beginning of next week :)

ElliotFreebairn committed:
Improve performance by reducing mask calls

From the flame graph, set data and recaculate contributed alot to CPU usage

Therefore, I chose to reduce the amount of recaculate calls by not removing them as member variables and thus only needing to caculate the masks when needed to

This does mean more code repitition however, it reduced execution time from 19 seconds to 13 seconds in one of my runs. The trade-off is worth it

Signed-off-by: elliotfreebairn <elliot.freebairn@collabora.com>

stbergmann commented:
Another approach might be to do costly operations only on demand, and then cache them, and then invalidate them when necessary.  (For example, `BMP::m_blurred_edge_mask` could be a `std::optional<std::vector<bool>>`, and `get_blurred_edge_mask` would check whether it's set, and if not calculate it first, and `recalculate_masks` would become some `invalidate_masks` that sets it to e an empty `optional` again.)

ElliotFreebairn commented:
> Another approach might be to do costly operations only on demand, and then cache them, and then invalidate them when necessary. (For example, `BMP::m_blurred_edge_mask` could be a `std::optional<std::vector<bool>>`, and `get_blurred_edge_mask` would check whether it's set, and if not calculate it first, and `recalculate_masks` would become some `invalidate_masks` that sets it to e an empty `optional` again.)

That's a really good idea, thank you

ElliotFreebairn committed:
Cache and invalidate masks only on demand

Instead of having to pass the masks, which were calculated outside of pixelbasher/bmp code, now are they are stored inside of the BMP

However, they are std::optional so that when `get_*_mask` is called, it checks whether the variable is empty, and if so calculates a new value

And now, when the set_data is called, it will invalidate the masks, making the std::optional empty

This approach keeps the previous performance improvements, but greatly reduces code duplication and betters the readbility

Signed-off-by: elliotfreebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
background values are cached & invalidated

Signed-off-by: elliotfreebairn <elliot.freebairn@collabora.com>

ElliotFreebairn commented:
This approach is definitely a lot better and best of all the performance improvements retained/even improved with the execution time hovering around 9.5 seconds

A quick question about one of my changes, in `Pixeblasher::compare_bmps()`, the BMP's were `const` but calling `get_*_masks`   change member variables therefore the objects being called upon couldn't be `const`

My solution was to not get rid of `const` but make the std::optional variables `mutable`, is that a good approach?

ElliotFreebairn committed:
Remove debugging code in pixelbasher test

Signed-off-by: elliotfreebairn <elliot.freebairn@collabora.com>

stbergmann commented:
> My solution was to not get rid of `const` but make the std::optional variables `mutable`, is that a good approach?

Yes, that's the solution of choice for such nominally-const behavior with caching data members.

ElliotFreebairn committed:
Script to produce flamegraph

Signed-off-by: elliotfreebairn <elliot.freebairn@collabora.com>

ElliotFreebairn commented:
I've created a very basic (and probably not good one right now) script which produces a flame graph for the program.

Should this be part of the public repo, or should I keep it local? Also, the script currently clones a GitHub repository (FlameGraph), would that be considered a bad approach?

stbergmann reviewed (commented):
(One thing I always kept forgetting to mention is that `PixelBasher` can be a namespace instead of a class, and you can get rid of the `PixelBasher` instance in `main.cpp`.)

stbergmann commented:
> I've created a very basic (and probably not good one right now) script which produces a flame graph for the program.
> 
> Should this be part of the public repo, or should I keep it local? Also, the script currently clones a GitHub repository (FlameGraph), would that be considered a bad approach?

I think submitting the script to the repo is a good idea.  And if it's just some script that lies around there for reference and when somebody wants to do new performance tests, but isn't routinely executed as part of making the project, then I think letting the script check out that FlameGraph repo is OK.

ElliotFreebairn committed:
Making pixelbasher a namespace

Signed-off-by: elliotfreebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Comments  to help the run flamegraph script

Signed-off-by: elliotfreebairn <elliot.freebairn@collabora.com>

ElliotFreebairn commented:
Flamegraph as an SVG attachment, rather than just a screenshot:

![flamegraph](https://github.com/user-attachments/assets/c235a244-a04f-478f-9035-5b3bf12a11b5)

stbergmann commented on src/pixelbasher.hpp:
```diff
@@ -19,17 +19,14 @@
+    PixelValues compare_pixel_regression(BMP& diff, PixelValues original, PixelValues current, PixelValues previous);
+    PixelValues compare_pixels(PixelValues original, PixelValues target, BMP &diff, int original_background_value, bool near_edge, bool vertical_edge, bool minor_differences);
+    PixelValues colour_pixel(Colour colour);
 };
```
(minor nitpick: there's an excess ; at the end of this line now)

ElliotFreebairn committed:
Using AVX2 intrinsics intersection_mask

This alone reduced `calculate_intersection_mask_simd` samples from 13% to 10.9%

Will start working on adding/replacing filtering and mask logic with AVX2

However, this the masks had to be converted to `uint8_t` from `bool` so that AVX2 can handle them

Signed-off-by: elliotfreebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Remove ;

Signed-off-by: elliotfreebairn <elliot.freebairn@collabora.com>

ElliotFreebairn commented:
Michael recommended in phab that I look into AVX2 intrinsics, so I spent all morning researching and seeing if I could make it work

The commit: `3a47f5b2ee833f3a80ceed0f012e32390e6f138b` contains a solution which replaces the `calculate_intersection_mask` scalar logic with the AVX2 SIMD implementation

It reduced `calculate_intersection_mask` sample percentage from **28.22 to 10.78%**.  And also the execution time went from: **11.1746 seconds to 7.0066 seconds**. There might have been some external differences, but hopefully the reductions show that implementing more of AVX2 code would be good

I have one little worry, which is that the flame graphs are very different and so is the execution time. I wouldn't think that just changing the intersection mask logic would make that differences. Maybe because all the masks are stored and calculated with `uint8_t`s not `bool` that added extra performance improvements anyways? 

Anyways, I am going to figure out how to check if the system can use AVX2, if not It can keep and use the previous scalar logic.

Here is the flamegraph pre-commit:
![old-flamegraph](https://github.com/user-attachments/assets/4c3aab8b-55a9-4108-9435-9b1bcff7045b)

flamegraph post-commit:
![flamegraph](https://github.com/user-attachments/assets/bf289943-6b7e-472b-9d89-71661beae7aa)

stbergmann commented:
Yeah, `std::vector<bool>::operator[]` is somewhat costly (esp. in a non-optimized build, cf. lack of any `-O...` in `makefile`'s `CXXFLAGS`), so it's probably no surprise that the original spent quite a chunk of `pixelbasher::compare_bmps` there.

(A "middle ground" between a bit/`bool`-wise `operator&&` and a 32-byte-wise `_mm256_and_si256` might be a 8-byte wise `uint64_t` `operator&`.  And maybe even trying out whether at high-enough optimization levels (`-O3`), the compiler is then smart enough to replace your loop with whatever SIMD-instructions it sees fit.)

ElliotFreebairn commented:
Wow, I added `-O3` to CXX flags, without any other changes, and It reduced execution time to 5 seconds. Thanks for pointing that out, I don't know if I would have found that otherwise

> (A "middle ground" between a bit/`bool`-wise `operator&&` and a 32-byte-wise `_mm256_and_si256` might be a 8-byte wise `uint64_t` `operator&`. And maybe even trying out whether at high-enough optimization levels (`-O3`), the compiler is then smart enough to replace your loop with whatever SIMD-instructions it sees fit

Are you saying that its best to go for `uint64_t` `operator&` as its less complicated and portable than the `_mm256_and_si256` for example, not having to `#include "immintrin.h"` ?

ElliotFreebairn committed:
Include optimization in makefile CXX Flags

Signed-off-by: elliotfreebairn <elliot.freebairn@collabora.com>

stbergmann commented:
> Wow, I added `-O3` to CXX flags, without any other changes, and It reduced execution time to 5 seconds. Thanks for pointing that out, I don't know if I would have found that otherwise

yeah, sorry I missed to bring that up earlier

> > (A "middle ground" between a bit/`bool`-wise `operator&&` and a 32-byte-wise `_mm256_and_si256` might be a 8-byte wise `uint64_t` `operator&`. And maybe even trying out whether at high-enough optimization levels (`-O3`), the compiler is then smart enough to replace your loop with whatever SIMD-instructions it sees fit
> 
> Are you saying that its best to go for `uint64_t` `operator&` as its less complicated and portable than the `_mm256_and_si256` for example, not having to `#include "immintrin.h"` ?

Only measurement can tell what's actually the best strategy overall.  (Note: with your current `std::vector<uint8_t>`, you are "wasting" 8 bits instead of just 1 bit per pixel, so the `_mm256_and_si256` will only operate on 32 pixels at a time.  If you kept using `std::vector<bool>`, which packs 8 `bool`s, i.e., 8 pixels, into a byte, that might be a bit more expensive at the places where you read/write individual elements via `operator[]`, but the `_mm256_and_si256` could then operate on 256 pixels in parallel, so that might well be faster overall.)

ElliotFreebairn committed:
Remove 'no_save_overlay' from performance testing

Signed-off-by: elliotfreebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Remove optimisations for performance testing

Signed-off-by: elliotfreebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Allow specific documents to be tested

Signed-off-by: elliotfreebairn <elliot.freebairn@collabora.com>

stbergmann commented on makefile:
```diff
@@ -1,12 +1,21 @@
+# Default target
+all: release
+
+debug: CXXFLAGS += -g -DDEBUG
```
there's no need to define a `DEBUG` macro if you don't use it (the `NDEBUG` macro below is a different story, as it is observed by the `assert.h`/`cassert` standard header)

ElliotFreebairn committed:
Force specifying extensions for flamegraphs

Signed-off-by: elliotfreebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Remove DEBUG macro

Signed-off-by: elliotfreebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Change file permissions to be owned by the user

Signed-off-by: elliotfreebairn <elliot.freebairn@collabora.com>

ElliotFreebairn commented:
@stbergmann Is there anything else obvious you could think I can work on/improve? Or even new things to try out :)

stbergmann commented on src/bmp.cpp:
```diff
@@ -558,23 +553,87 @@ int BMP::calculate_colour_count(const BMP& base, Colour to_compare)
+    std::vector<uint8_t> intersection_mask(width * height, 0);
+    for (int i = 0; i < width * height; i++)
+    {
+        intersection_mask[i] = original_edge_map[i] && target_edge_map[i];
```
I had accidentally commented about this at <https://github.com/CollaboraOnline/mso-test/commit/3a47f5b2ee833f3a80ceed0f012e32390e6f138b#r164625979>, where you presumably didn't notice it (Github PR review really is a pain):  "it might even be cheaper then to replace short-circuiting `operator&&` with bitwise `operator&`"

ElliotFreebairn replied:
Yeah that makes a lot of sense, thank you

stbergmann commented on src/bmp.hpp:
```diff
@@ -15,7 +15,9 @@
 #include <iostream>
+#include <optional>
 #include <vector>
+#include "immintrin.h"
```
do you actually need this here, or only in `bmp.cpp`?  (and by convention, `#include <...>` style is used for all kinds of include files that come "from the implementation", not just for the standard headers)

stbergmann commented on src/pixelbasher.cpp:
```diff
@@ -15,16 +15,21 @@
-    std::vector<bool> target_filtered_vertical_edges = target.get_filtered_vertical_edge_mask();
+
+
+    // const std::vector<uint8_t> intersection_mask = BMP::calculate_intersection_mask(original, target);
```
drop the commented-out code and its preceding newlines?

stbergmann commented:
> @stbergmann Is there anything else obvious you could think I can work on/improve? Or even new things to try out :)

did you look into <https://github.com/CollaboraOnline/mso-test/pull/25#issuecomment-3224355871> "Only measurement can tell what's actually the best strategy overall"?

ElliotFreebairn commented:
> did you look into <https://github.com/CollaboraOnline/mso-test/pull/25#issuecomment-3224355871> "Only measurement can tell what's actually the best strategy overall"?

Oh I'm an idiot, I had a look yesterday but got distracted and forgot. I'll look into that today

ElliotFreebairn committed:
Only need AVX2 header inside bmp.cpp

Signed-off-by: elliotfreebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Remove comment and  newlines

Signed-off-by: elliotfreebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Replace logic && and with bit wise operator &

Signed-off-by: elliotfreebairn <elliot.freebairn@collabora.com>

ElliotFreebairn commented:
> > @stbergmann Is there anything else obvious you could think I can work on/improve? Or even new things to try out :)
> 
> did you look into [#25 (comment)](https://github.com/CollaboraOnline/mso-test/pull/25#issuecomment-3224355871) "Only measurement can tell what's actually the best strategy overall"?

I've just had a look, so it seems that it's seems that you can't load in a `vector<bool>` with `_mm256_loadu_si256`. I will look into switching the vectors to `uint64_t` as `_mm256_loadu_si256` is able to load that in

ElliotFreebairn committed:
Masks are stored in uin64_t vectors not uint8_t

Signed-off-by: elliotfreebairn <elliot.freebairn@collabora.com>

ElliotFreebairn commented:
I've just spent quite a bit of time implementing switching the masks from being stored as `uint8_t` to `uint64_t`, but the performance increases were negligible or even slightly worse sometimes

Do you think this is caused by the extra work being done by `get_bit` and `set_bit`, or Is this change worth it?

If It's not worth it, I can always go back to the previous version

stbergmann commented:
Yeah, if packing the bits didn't bring a real performance benefit (which indeed is likely if the individual-bit getting/setting functionality dominates the bulk functionality in the actual use cases), I'd not bother with the additional code complexity that would bring, and return to the simpler scheme.

ElliotFreebairn committed:
Revert to "3cb2c8712570e017d0db9167c249e8ebe891e90b"

Signed-off-by: elliotfreebairn <elliot.freebairn@collabora.com>
